src/refda/exec.c

```diff
@@ -375,91 +376,260 @@ void *refda_exec_worker(void *arg)
+        while (!refda_timeline_end_p(tl_it))
+        {
+            const refda_timeline_event_t *next = refda_timeline_cref(tl_it);
+            if (timespec_gt(next->ts, nowtime))
             {
-                const refda_timeline_event_t *next = refda_timeline_cref(tl_it);
-                if (timespec_gt(next->ts, nowtime))
+                break;
+            }
+
+            CACE_LOG_DEBUG("running deferred callback");
+            switch (next->purpose)
+            {
+                case REFDA_TIMELINE_CTRL:
                 {
+                    {
+                        refda_ctrl_exec_ctx_t ctx;
+                        refda_ctrl_exec_ctx_init(&ctx, next->exec.item);
+                        (next->exec.callback)(&ctx);
+                        refda_ctrl_exec_ctx_deinit(&ctx);
+                    }
+                    if (!atomic_load(&(next->exec.item->waiting)))
+                    {
+                        refda_exec_ctrl_finish(next->exec.item);
+                    }
                     break;
                 }
-
-                CACE_LOG_DEBUG("running deferred callback");
+                case REFDA_TIMELINE_TBR:
                 {
-                    refda_ctrl_exec_ctx_t ctx;
-                    refda_ctrl_exec_ctx_init(&ctx, next->item);
-                    (next->callback)(&ctx);
-                    refda_ctrl_exec_ctx_deinit(&ctx);
+                    (next->tbr.callback)(next->tbr.agent, next->tbr.tbr);
+                    break;
                 }
-                if (!atomic_load(&(next->item->waiting)))
+                default:
                 {
-                    refda_exec_ctrl_finish(next->item);
+                    CACE_LOG_ERR("Unknown type of deferred callback %d", next->purpose);
                 }
-
-                refda_timeline_remove(agent->exec_timeline, tl_it);
             }
+
+            refda_timeline_remove(agent->exec_timeline, tl_it);
         }
-        else
+    }
+    else
+    {
+        CACE_LOG_DEBUG("waiting for exec event");
+        sem_wait(&(agent->execs_sem));
+    }
+
+    // execs queue may still be empty if deferred callbacks were run
+    if (refda_msgdata_queue_pop(&item, agent->execs))
+    {
+        // sentinel for end-of-input
+        const bool at_end = cace_ari_is_undefined(&(item.value));
+        if (!at_end)
+        {
+            refda_exec_exp_execset(agent, &item);
+        }
+        refda_msgdata_deinit(&item);
+        if (at_end && refda_timeline_empty_p(agent->exec_timeline))
         {
-            CACE_LOG_DEBUG("waiting for exec event");
-            sem_wait(&(agent->execs_sem));
+            CACE_LOG_INFO("Got undefined exec, stopping");
+
+            // flush the input queue but keep the daemon running
+            refda_msgdata_t undef;
+            refda_msgdata_init(&undef);
+            refda_msgdata_queue_push_move(agent->rptgs, &undef);
+            sem_post(&(agent->rptgs_sem));
+
+            return;
```

Comment:
This is the loop exit condition.

Reply:
Great catch! Thanks for the assist on this one, just wasn't seeing it during my subsequent read through the code.